src/audio/pipewire/SDL_pipewire.c

```diff
@@ -0,0 +1,1088 @@
+    return;
+  }
+
+  spa_buf = pw_buf->buffer;
+
+  if (spa_buf->datas[0].data == NULL) {
+    return;
+  }
+
+  /*
+   * If the device is disabled, write silence to the stream buffer
+   * and run the callback with the work buffer to keep the callback
+   * firing regularly in case the audio is being used as a timer.
+   */
+  if (SDL_AtomicGet(&this->enabled)) {
+    dst = spa_buf->datas[0].data;
+  } else {
+    dst = this->work_buffer;
+    SDL_memset(spa_buf->datas[0].data, this->spec.silence, this->spec.size);
+  }
+
+  if (!SDL_AtomicGet(&this->paused)) {
+    if (!this->stream) {
+      SDL_LockMutex(this->mixer_lock);
+      this->callbackspec.callback(this->spec.userdata, dst, this->callbackspec.size);
+      SDL_UnlockMutex(this->mixer_lock);
+    } else {
+      int got;
+
+      /* Fire the callback until we have enough to fill a buffer */
+      while (SDL_AudioStreamAvailable(this->stream) < this->spec.size) {
+        SDL_LockMutex(this->mixer_lock);
+        (*this->callbackspec.callback)(this->callbackspec.userdata, this->work_buffer, this->callbackspec.size);
+        SDL_UnlockMutex(this->mixer_lock);
+
+        SDL_AudioStreamPut(this->stream, this->work_buffer, this->callbackspec.size);
+      }
+
+      got = SDL_AudioStreamGet(this->stream, dst, this->spec.size);
+      SDL_assert(got == this->spec.size);
+    }
+  } else if (dst != this->work_buffer) {
+    SDL_memset(dst, this->spec.silence, this->spec.size);
+  }
+
+  spa_buf->datas[0].chunk->offset = 0;
+  spa_buf->datas[0].chunk->stride = this->hidden->stride;
+  spa_buf->datas[0].chunk->size   = this->spec.size;
+
+  PIPEWIRE_pw_stream_queue_buffer(stream, pw_buf);
+}
+
+static void
+input_callback(void *data)
+{
+  struct pw_buffer * pw_buf;
+  struct spa_buffer *spa_buf;
+  Uint8 *            src;
+  _THIS                    = (SDL_AudioDevice *)data;
+  struct pw_stream *stream = this->hidden->stream;
+
+  /* Shutting down, don't do anything */
+  if (SDL_AtomicGet(&this->shutdown)) {
+    return;
+  }
+
+  pw_buf = PIPEWIRE_pw_stream_dequeue_buffer(stream);
+  if (!pw_buf) {
+    return;
+  }
+
+  spa_buf = pw_buf->buffer;
+
+  if ((src = (Uint8 *)spa_buf->datas[0].data) == NULL) {
+    return;
+  }
+
+  if (SDL_AtomicGet(&this->enabled)) {
+    /* The first packet might be undersized, so pad the start with silence */
+    if (spa_buf->datas[0].chunk->size < this->spec.size) {
+      const Uint32 diff = this->spec.size - spa_buf->datas[0].chunk->size;
+      SDL_memmove(src + diff, src, diff);
+      SDL_memset(src, this->spec.silence, diff);
+    }
+  } else { /* Zero the buffer if the stream is disabled. */
+    SDL_memset(src, this->spec.silence, this->spec.size);
+  }
+
+  if (!this->stream) {
+    /* No intermediate stream, call the application callback directly */
+    if (!SDL_AtomicGet(&this->paused)) {
+      SDL_LockMutex(this->mixer_lock);
+      this->callbackspec.callback(this->callbackspec.userdata, src, this->callbackspec.size);
```

Comment:
Recording like this won't work. The latency specified in PIPEWIRE_NODE_LATENCY only guarantees a minimal latency to or app. The actual latency might be lower depending on the graph. For example, just opening pavucontrol or recording from the same source from a different app that request a lower latency will lower the latency of the graph enough, that chunk->size < this->spec.size is always true and we get broken recording. We probably need to put an additional ringbuffer in between.

Reply:
Ah, I see.  Rewriting it to always buffer the audio is simple enough, and it's just a bit of overhead from copying some bytes around.  I wanted to do some work on that function anyways since I realized I'm ignoring the n_datas member of spa_buf.  Not an issue in the output callback as we only use one buffer, but there is no guarantee of that when dealing with incoming audio.

Reply:
I might be wrong, but i''m pretty sure the n_datas member will always be 1 when you request an interleaved audio format, so it should be fine. It would be different with planar audio formats, but since SDL doesn't support those, we don't have to care.

Reply:
That's what it sounds like in the documentation, although it's a little vague.  
> Each data block is, for example, a video plane or an audio channel. There are n_datas of those blocks.

This is where it would be nice to have a detailed spec stating something like
> If the audio sample format is interleaved, n_datas will always be 1

Probably not worth worrying about though.